File: Stack/503.Next-Greater-Element-II/503.Next-Greater-Element-II.ts
function nextGreaterElements(nums: number[]): number[] {
  const ret = new Array(nums.length).fill(-1);
  const stack = [];
  for (let i = 0; i < nums.length * 2; i++) {
    const idx = i % nums.length;
    while (stack.length !== 0 && nums[idx] > nums[stack[stack.length - 1]]) {
      const index = stack.pop();
      ret[index] = nums[idx];
    }
    stack.push(idx);
  }
  return ret;
}

File: Stack/503.Next-Greater-Element-II/503.Next-Greater-Element-II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        result = [-1] * len(nums)

        stack = [0]
        for i in range(1, len(nums) * 2):
            i = i % len(nums)
            while len(stack) and nums[i] > nums[stack[-1]]:
                index = stack.pop()
                result[index] = nums[i]
            stack.append(i)
        return result
                
